---
title: "Concordance"
author: "Steven N Hart"
date: "09/14/2015"
output: html_document
---

```{r, echo=FALSE}
library("ggplot2")
library(knitr)
library("gridExtra")
#setwd("/data5/bsi/epibreast/m087494.couch/CGSL/Verification/TEST")

#Usually we will do each one of these types of tests.  What this part of the script does it is looks for folders that start with the test type and reads in those data.  It needs to skip any image file that might already exists.
prefixes=c("InterAssay","IntraFlowcell","IntraLibrary","LowInput","Reproducibility")
Data<-NULL
for (prefix in prefixes){
  analyses=dir()[grep(prefix,dir())]
  analyses<-analyses[grep("png",analyses,invert=T)]
  
  for (analysis in analyses){
    print(analysis)
    tmp<-read.table(paste(analysis,"/concordance.out",sep=""),na.strings = ".",sep="\t")
    names(tmp)<-c("name1","name2","Concordance","Type","Size","CONCORDANCE","COUNT1","COUNT2","CONCORDANCE2","COUNT3","COUNT4","UNUSED")
    tmp$CONCORDANCE[which(tmp$CONCORDANCE>1)]<-1
    tmp$CONCORDANCE2[which(tmp$CONCORDANCE2>1)]<-1
    tmp$Prefix=prefix
    Data<-rbind(Data,tmp)
  }
}
##################################################
#Use overlap instead of exact genotype
#Data$CONCORDANCE<-Data$CONCORDANCE2
#Data$COUNT1<-Data$COUNT3
#Data$COUNT2<-Data$COUNT4
##################################################

#label probands
Data$Proband="LowInput"
Data$Proband[grep("ProbA",Data$name1)]<-"ProbA"
Data$Proband[grep("MomA",Data$name1)]<-"ProbA"
Data$Proband[grep("DadA",Data$name1)]<-"ProbA"
Data$Proband[grep("ProbB",Data$name1)]<-"ProbB"
Data$Proband[grep("MomB",Data$name1)]<-"ProbB"
Data$Proband[grep("DadB",Data$name1)]<-"ProbB"
Data$Proband[grep("ProbC",Data$name1)]<-"ProbC"
Data$Proband[grep("MomC",Data$name1)]<-"ProbC"
Data$Proband[grep("DadC",Data$name1)]<-"ProbC"

#Label standards
Probands<-c("ProbA","ProbB","ProbC")
Types<-c("SNPS","INS","DEL")

#Create a backup in case I delete the original
Data.backup<-Data
#options(warn=-1)
Table1=NULL
OVERALL_PLOTS=NULL
for (ASSAY in prefixes){
  for (TYPE in Types){
    OVERALL=NULL
    replicate=0
    for (PROBAND in Probands){
      #subset the data
      if(PROBAND!="ProbA" & ASSAY=="LowInput"){ break}
      if(ASSAY=="Reproducibility" & replicate > 0){ break}
      replicate=replicate+1
      tmp<-Data[which(Data$Prefix == ASSAY & Data$Proband == PROBAND & Data$Type == TYPE),]
      tmp<-unique(tmp)
      MEAN_CONCORDANCE=tapply(tmp$CONCORDANCE,tmp$Size,mean)
      MIN_CONCORDANCE=tapply(tmp$CONCORDANCE,tmp$Size,min)
      MAX_CONCORDANCE=tapply(tmp$CONCORDANCE,tmp$Size,max)
      Size<-names(MEAN_CONCORDANCE)
      TOTAL_EVENTS=tapply(tmp$COUNT1,tmp$Size,mean)
      NUM_EVENTS=round(tapply(tmp$COUNT1*tmp$CONCORDANCE,tmp$Size,mean))
      
  ###
    tmp$NUM_EVENTS=NUM_EVENTS
    tmp$NUM_EVENTS[which(is.na(tmp$NUM_EVENTS))]<-0
    tmp$TOTAL_EVENTS=tapply(tmp$COUNT1,tmp$Size,mean)
    tmp$TOTAL_EVENTS[which(is.na(tmp$TOTAL_EVENTS))]<-0
    
    LEGEND=sum(tmp$NUM_EVENTS)/sum(tmp$TOTAL_EVENTS)
    OVERALL=round(LEGEND,digits=3)
      #Get number of comparisons made and double it
    NAMES=data.frame(table(tmp$name1,tmp$name2))
    NAMES=NAMES[which(NAMES$Freq>0),]
    NSAMPLES=length(unique(c(as.character(NAMES$Var1),as.character(NAMES$Var2))))
    NCOMPARISONS=dim(NAMES)[1]    
  ###      
      
      res=data.frame(Size,MEAN_CONCORDANCE,MIN_CONCORDANCE,MAX_CONCORDANCE,ASSAY,PROBAND,TYPE, replicate,NUM_EVENTS,TOTAL_EVENTS,OVERALL,NSAMPLES,NCOMPARISONS)
      OVERALL_PLOTS<-rbind(OVERALL_PLOTS,res)
  }
}
}
tmp<-unique(OVERALL_PLOTS[,c(5,6,7,11)])
OVERALL_TABLE=aggregate(OVERALL ~TYPE + ASSAY,mean,data=tmp)
PROBAND_TABLE=aggregate(OVERALL ~TYPE + ASSAY + PROBAND,mean,data=tmp)
ASSAY_TABLE=aggregate(MEAN_CONCORDANCE ~ Size + ASSAY + TYPE,data=OVERALL_PLOTS, mean)
tmp<-unique(OVERALL_PLOTS[,c(1,2,5,6,7)])
SIZE_ASSAY_TABLE=aggregate(MEAN_CONCORDANCE ~TYPE + ASSAY + Size,mean,data=tmp)

tmp<-Data[which(Data$Prefix==prefixes[1] & Data$Proband==Probands[1]) ,c("Size","Type","COUNT1")]
tmp<-unique(tmp)
tmp=na.omit(tmp)
EVENTS_TABLE=aggregate(COUNT1 ~ Size + Type, mean,data=tmp)

write.table(OVERALL_TABLE,file="OVERALL_TABLE.tsv",sep = "\t",row.names = F)
write.table(PROBAND_TABLE,file="PROBAND_TABLE.tsv",sep = "\t",row.names = F)
write.table(SIZE_ASSAY_TABLE,file="SIZE_ASSAY_TABLE.tsv",sep = "\t",row.names = F)
write.table(EVENTS_TABLE,file="EVENTS_TABLE.tsv",sep = "\t",row.names = F)

```


Now I can make the aggregated plots
```{r echo=FALSE}
#Make each plot
i=0
for (ASSAY in prefixes){
  tmp<-SIZE_ASSAY_TABLE[which(SIZE_ASSAY_TABLE$ASSAY==ASSAY),]
  tmp$Size=as.numeric(tmp$Size)
  LEGEND=OVERALL_TABLE[which(OVERALL_TABLE$ASSAY == ASSAY),]
  LEGEND$OVERALL=round(LEGEND$OVERALL,digits = 3)
  p=NULL
    p<- ggplot(tmp,aes(x=Size,y=MEAN_CONCORDANCE,order=TYPE,group=TYPE,colour=TYPE))+ylim(0,1)
    p<- p+labs(title =ASSAY)+scale_color_manual( values = c("black", "red","blue"))+geom_point()+geom_line()
    p<-p+theme(legend.position=c(0.1,0.1))
  p<-p+scale_color_manual("Overall",labels =paste(LEGEND$TYPE,LEGEND$OVERALL,sep=" = "), values = c("black", "red","blue"))
    ggsave(p,file=paste(ASSAY,"png",sep="."))
  i=i+1
  if(i==1){a=p+theme(legend.position=c(0.2,0.2))}
  if(i==2){b=p+theme(legend.position=c(0.2,0.2))}
  if(i==3){c=p+theme(legend.position=c(0.2,0.2))}
  if(i==4){d=p+theme(legend.position=c(0.2,0.2))}
  if(i==5){e=p+theme(legend.position=c(0.2,0.2))}
}


f=ggplot(EVENTS_TABLE,aes(x=as.numeric(Size),y=COUNT1,group=Type,fill=Type))+geom_bar(stat="identity",position="dodge")+theme(legend.position=c(0.8,0.8))+ylim(0,1500)+labs(title ="#Events")

png(file="OverallConcordance.png",units="in",res=300,height=8,width=16)
grid.arrange(a,b,c,d,e,f,ncol=3)
dev.off()
```
Now I can make the Separated plots
```{r echo=FALSE}
#Make each plot
i=0
for (ASSAY in prefixes){
  for(PROBAND in Probands){
  tmp<-OVERALL_PLOTS[which(OVERALL_PLOTS$ASSAY==ASSAY & OVERALL_PLOTS$PROBAND==PROBAND),]
  tmp$Size=as.numeric(tmp$Size)
  LEGEND=PROBAND_TABLE[which(PROBAND_TABLE$ASSAY == ASSAY & PROBAND_TABLE$PROBAND==PROBAND),]
  p=NULL
  p<- ggplot(tmp,aes(x=Size,y=MEAN_CONCORDANCE,order=TYPE,group=TYPE,colour=TYPE))+ylim(0,1)
    p<- p+labs(title =paste(PROBAND,ASSAY))+scale_color_manual( values = c("black", "red","blue"))+geom_point()+geom_line()
    p<-p+theme(legend.position=c(0.1,0.1))
  p<-p+scale_color_manual("Overall",labels =paste(LEGEND$TYPE,LEGEND$OVERALL,sep=" = "), values = c("black", "red","blue"))
  p
    ggsave(p,file=paste(ASSAY,PROBAND,"png",sep="."))
  }
}
```
